presto-parquet/src/main/java/com/facebook/presto/parquet/reader/ParquetReader.java

```diff
@@ -408,4 +515,77 @@ private static Block rewriteLongArrayBlock(LongArrayBlock longArrayBlock, Type t
 
         return newBlockBuilder.build();
     }
+
+    private static <T> List<T> listWithNulls(int size)
+    {
+        return Stream.generate(() -> (T) null).limit(size).collect(Collectors.toCollection(ArrayList<T>::new));
+    }
+
+    private RowRanges getRowRanges(int blockIndex)
+    {
+        assert filter != null;
+
+        RowRanges rowRanges = blockRowRanges.get(blockIndex);
+        if (rowRanges == null) {
+            rowRanges = ColumnIndexFilter.calculateRowRanges(FilterCompat.get(filter), blockIndexStores.get(blockIndex),
+                    paths.keySet(), blocks.get(blockIndex).getRowCount());
+            blockRowRanges.set(blockIndex, rowRanges);
+        }
+        return rowRanges;
+    }
+
+    /**
+     * Describes a list of consecutive parts to be read at once. A consecutive part may contain whole column chunks or
+     * only parts of them (some pages).
+     */
+    private class PageRanges
+    {
+        private final long offset;
+        private int length;
+
+        /**
+         * @param offset where the first chunk starts
+         */
+        public PageRanges(long offset)
+        {
+            this.offset = offset;
+            this.length = 0;
+        }
+
+        public void extendLength(int length)
+        {
+            this.length += length;
+        }
+
+        public long getOffset()
+        {
+            return offset;
+        }
+
+        public int getLength()
+        {
+            return length;
+        }
+
+        public long endPos()
+        {
+            return offset + length;
+        }
+    }
+
+    private List<PageRanges> concatRanges(List<OffsetRange> offsetRanges)
+    {
+        List<PageRanges> pageRanges = new ArrayList<>();
+        PageRanges currentParts = null;
+        for (OffsetRange range : offsetRanges) {
+            long startPosition = range.getOffset();
+            // first part or not consecutive => new list
+            if (currentParts == null || currentParts.endPos() != startPosition) {
+                currentParts = new PageRanges(startPosition);
+            }
+            pageRanges.add(currentParts);
+            currentParts.extendLength((int) range.getLength());
```

Comment:
we are casting `OffsetRange` length to `int`, shall we use `OffsetRange` directly?

Reply:
Sounds good